fatoptimizer/inline.py

```diff
@@ -1,35 +1,128 @@
+
+def locate_kwarg(funcdef, name):
+    '''Get the index of an argument of funcdef by name.'''
+    for idx, arg in enumerate(funcdef.args.args):
+        if arg.arg == name:
+            return idx
+    raise ValueError('argument %r not found' % name)
+
+class RenameVisitor(NodeTransformer):
+    def __init__(self, callsite, inlinable, actual_pos_args):
+        assert callsite.starargs is None
+        assert callsite.kwargs is None
+        assert inlinable.args.vararg is None
+        assert inlinable.args.kwonlyargs == []
+        assert inlinable.args.kw_defaults == []
+        assert inlinable.args.kwarg is None
+        assert inlinable.args.defaults == []
+
+        # Mapping from name in callee to node in caller
+        self.remapping = {}
+        for formal, actual in zip(inlinable.args.args, actual_pos_args):
+            self.remapping[formal.arg] = actual
+
+    def visit_Name(self, node):
+        if node.id in self.remapping:
+            return self.remapping[node.id]
+        return node
+
+class Expansion:
+    '''Information about a callsite that's a candidate for inlining, giving
+    the funcdef, and the actual positional arguments (having
+    resolved any keyword arguments.'''
+    def __init__(self, funcdef, actual_pos_args):
+        self.funcdef = funcdef
+        self.actual_pos_args = actual_pos_args
 
 class InlineSubstitution(OptimizerStep):
     """Function call inlining."""
 
-    def can_inline(self, node):
-        '''Given a Call node, determine whether we should inline
-        the callee.  If so, return the callee FunctionDef, otherwise
-        return None'''
+    def can_inline(self, callsite):
+        '''Given a Call callsite, determine whether we should inline
+        the callee.  If so, return an Expansion instance, otherwise
+        return None.'''
         # TODO: size criteria?
-        # TODO: don't do it if either uses locals()
         # TODO: don't do it for recursive functions
-        if not isinstance(node.func, ast.Name):
+        if not isinstance(callsite.func, ast.Name):
             return None
         from .namespace import _fndefs
-        if node.func.id not in _fndefs:
+        if callsite.func.id not in _fndefs:
             return None
-        candidate = _fndefs[node.func.id]
+        candidate = _fndefs[callsite.func.id]
 
+        # For now, only support simple positional arguments
+        # and keyword arguments
+        if callsite.starargs:
+            return False
+        if callsite.kwargs:
+            return False
+        if candidate.args.vararg:
+            return False
+        if candidate.args.kwonlyargs:
+            return False
+        if candidate.args.kw_defaults:
+            return False
+        if candidate.args.kwarg:
+            return False
+        if candidate.args.defaults:
+            return False
+
+        # Attempt to match up the calling convention at the callsite
+        # with the candidate funcdef
+        if 0:
```

Comment:
please don't push code with print(), even if it's dead code. Keep them as local changes.